src/package.coffee

```diff
@@ -318,6 +320,21 @@ class Package
           @activationDisposables.add @packageManager.serviceHub.consume(name, version, @mainModule[methodName].bind(@mainModule))
     return
 
+  registerUrlHandler: ->
+    handlerConfig = @getUrlHandler()
+    if methodName = handlerConfig?.method
+      @urlHandlerSubscription = @packageManager.registerUrlHandlerForPackage @name, (args...) =>
+        @handleUrl(methodName, args)
+
+  unregisterUrlHandler: ->
+    @urlHandlerSubscription?.dispose()
+
+  handleUrl: (methodName, args) ->
+    @activate().then =>
+      @mainModule[methodName]?.apply(@mainModule, args)
+    unless @mainActivated
+      @activateNow()
```

Comment:
Not _100%_ sure this is the correct way to set things up, but it seems to work in all cases (including deferred loading).

Reply:
Would it be better to activate the package before handling the URL in case the handler depends on any state that gets established during activation?

Reply:
@nathansobo Unless I'm misreading the code, that's what's happening:

1. `activate()` creates the activation promise (a `Promise.all` of a few internal promises)
2. `activateNow()` resolves the activation promise
3. The `then` hook runs and calls the URL handler method

Reply:
Oh of course, the `then` is deferred. Carry on.

Reply:
Would it be possible to swap the statement order?